### PR TITLE
11205 - Improved Polygon Editor (cut/paste vertices, delete polygon) & Improved Region Editor (drag-at-edge works again, move regions inside of zone) 

### PR DESCRIPTION
**Pull request by Cattlesquat**

### CODE REVIEW

File: vassal-app/src/main/java/VASSAL/build/module/map/boardPicker/board/RegionGrid.java [hunk 1]
```diff
@@ -897,9 +918,20 @@ public void mouseClicked(MouseEvent e) {
...
 
+    protected void scrollToZone() {
+      final Zone z = grid.getZone();
+      final Rectangle polyBounds = z.getBounds();
+      final Point polyCenter = new Point(polyBounds.x + polyBounds.width / 2,
+        polyBounds.y + polyBounds.height / 2);
+      final Rectangle rect = new Rectangle(polyCenter);
+      scroll.getViewport().scrollRectToVisible(rect);
+    }
+
```
Comment (uckelman) [inline vassal-app/src/main/java/VASSAL/build/module/map/boardPicker/board/RegionGrid.java:929]: I think what you need is this:

```
final Rectangle rect = z.getBoard().getMap().mapToComponent(z.getBounds());
rect.setBounds(rect.x + rect.width / 2, rect.y + rect.height / 2, 0, 0);
scroll.getViewport().scrollRectToVisible(rect);
```

File: vassal-app/src/main/java/VASSAL/build/module/map/boardPicker/board/RegionGrid.java [hunk 2]
```diff
@@ -973,12 +1089,38 @@ else if (command.equals(DELETE_REGION)) {
...
+        final Zone zone = grid.getZone();
+        if (zone != null) {
+          final Polygon polygon = zone.getPolygon();
+          if ((polygon != null) && (polygon.npoints > 0)) {
+            final Rectangle rect = polygon.getBounds();
+            int num = 0;
+            for (final Region r : selectedRegions) {
+              final int offX = (num > 0) ? ((num - 1) % 3) - 1 : 0;
+              final int offY = ((num % 9) < 3) ? 0 : ((num % 9) < 6) ? -1 : 1;
+              r.setOrigin(new Point((int)rect.getCenterX() + offX * 10, (int)rect.getCenterY() + offY * 10));
```
Comment (uckelman) [inline vassal-app/src/main/java/VASSAL/build/module/map/boardPicker/board/RegionGrid.java:1105]: Woah, what does this do?

**Comment by Cattlesquat** [comment]
Mainly it takes errant points that are outside the zone and moves them into the zone “near the center”. The bonus frippery is if you move more than one point in a batch it makes some effort not to have them all land on same pixel (makes a little grid).

**Comment by uckelman** [comment]
How'd you arrive at the arithmetic for that?

**Pull request merged**